lucene/core/src/java/org/apache/lucene/index/StandardDirectoryReader.java

```diff
@@ -39,33 +40,47 @@
 
   final IndexWriter writer;
   final SegmentInfos segmentInfos;
+  private final Comparator<LeafReader> leafSorter;
   private final boolean applyAllDeletes;
   private final boolean writeAllDeletes;
 
-  /** called only from static open() methods */
+  /** package private constructor, called only from static open() methods. */
   StandardDirectoryReader(
       Directory directory,
       LeafReader[] readers,
       IndexWriter writer,
       SegmentInfos sis,
+      Comparator<LeafReader> leafSorter,
       boolean applyAllDeletes,
       boolean writeAllDeletes)
       throws IOException {
-    super(directory, readers);
+    super(directory, sortLeaves(readers, leafSorter));
```

Comment:
I wonder if the leafSorter should be declared and executed in the base class `BaseCompositeReader` ?
That would expose the feature explicitly to `MultiReader` and friends.